eval/eval.go

```diff
@@ -193,12 +195,27 @@ type evalContext struct {
 	root      *expr  // the root expression
 	base      *value // the base value
 
-	rotateDocPaths map[string]bool // the subset of document paths to invoke rotation for when rotating. if empty, all rotators will be invoked.
-	patchOutputs   []*Patch        // updated rotation state generated during evaluation, to be written back to the environment definition
+	rotateDocPaths  map[string]bool   // the subset of document paths to invoke rotation for when rotating. if empty, all rotators will be invoked.
+	patchOutputs    []*Patch          // updated rotation state generated during evaluation, to be written back to the environment definition
+	rotationResults []*RotationResult // list of results of rotations performed
 
 	diags syntax.Diagnostics // diagnostics generated during evaluation
 }
 
+type RotationResult struct {
+	Path   string
+	Status RotationStatus
+	Error  string
```

Comment:
Instead of an error string, let's return a slice of diagnostics. That should be a bit more future-proof.